src/main/java/com/metamx/common/CompressionUtils.java

```diff
@@ -0,0 +1,326 @@
+  {
+    if (!directory.isDirectory()) {
+      throw new IOException(String.format("directory[%s] is not a directory", directory));
+    }
+    final File[] files = directory.listFiles();
+
+    long totalSize = 0;
+    try (final ZipOutputStream zipOut = new ZipOutputStream(out)) {
+      for (File file : files) {
+        log.info("Adding file[%s] with size[%,d].  Total size so far[%,d]", file, file.length(), totalSize);
+        if (file.length() >= Integer.MAX_VALUE) {
+          zipOut.finish();
+          throw new IOException(String.format("file[%s] too large [%,d]", file, file.length()));
+        }
+        zipOut.putNextEntry(new ZipEntry(file.getName()));
+        totalSize += Files.asByteSource(file).copyTo(zipOut);
+      }
+      zipOut.closeEntry();
+    }
+
+    return totalSize;
+  }
+
+  public static FileUtils.FileCopyResult unzip(
+      final ByteSource inputStreamSupplier,
+      final File outDir,
+      final Predicate<Throwable> shouldRetry,
+      Boolean cacheLocally
+  ) throws IOException
+  {
+    if (!cacheLocally) {
+      return unzip(inputStreamSupplier.openStream(), outDir);
+    }
+    final File tmpFile = File.createTempFile("compressionUtilZipCache", ZIP_SUFFIX);
+    try {
+      FileUtils.FileCopyResult copyResult = FileUtils.retryCopy(
+          inputStreamSupplier,
+          shouldRetry,
+          tmpFile,
+          DEFAULT_RETRY_COUNT
+      );
+      return unzip(tmpFile, outDir);
+    }
+    finally {
+      if (!tmpFile.delete()) {
+        log.warn("Could not delete zip cache at [%s]", tmpFile.getAbsolutePath());
+      }
+    }
+  }
+
+  public static FileUtils.FileCopyResult unzip(
+      final ByteSource inputStreamSupplier,
+      final File outDir,
+      Boolean cacheLocally
+  ) throws IOException
+  {
+    return unzip(inputStreamSupplier, outDir, Predicates.<Throwable>alwaysTrue(), cacheLocally);
+  }
+
+  public static FileUtils.FileCopyResult unzip(final File pulledFile, final File outDir) throws IOException
+  {
+    if (!(outDir.exists() && outDir.isDirectory())) {
+      throw new ISE("outDir[%s] must exist and be a directory", outDir);
+    }
+    log.info("Unzipping file[%s] to [%s]", pulledFile, outDir);
+    final ArrayList<FileUtils.FileCopyResult> children = new ArrayList<>();
+    try (final ZipFile zipFile = new ZipFile(pulledFile)) {
+      final Enumeration<? extends ZipEntry> enumeration = zipFile.entries();
+      while (enumeration.hasMoreElements()) {
+        final ZipEntry entry = enumeration.nextElement();
+        children.add(
+            FileUtils.retryCopy(
+                new ByteSource()
+                {
+                  @Override
+                  public InputStream openStream() throws IOException
+                  {
+                    return new BufferedInputStream(zipFile.getInputStream(entry));
+                  }
+                },
+                FileUtils.MD5ERROR,
```

Comment:
I'm not sure what we are trying to guard against with retries here. When trying to retry a read from file, if the checksums don't match the first time, it's unlikely they will match the second time, no? Otherwise, this would indicate memory corruption issues, in which case we have much bigger problems.